File: data_analysis/findShortestPath.cpp
#include <iostream>
#include <string>
#include <array>
#include "sqlite3.h"
#include <chrono>
#include <unordered_map>
#include <unordered_set>
#include <vector>
#include <queue>
#include <deque>
#include "../handlers/dbHandler.h"
#include <nlohmann/json.hpp>
#include <fstream>
#include "findShortestPath.h"



using namespace std;


void writeToJSON(string artist_id, unordered_map<int, int> pathLengths, int execTime){
    try{
        const string fileName = "data/1000_random_artists_cpp.json";
    
        ifstream inFile(fileName);
        if (!inFile.is_open()) {
            throw std::runtime_error("Could not open input file.");
        }
        nlohmann::json jsonData;
        inFile >> jsonData; // Parse JSON file content into jsonData
        inFile.close();

        for (int i = 0; i < jsonData.size(); i++){
            if (jsonData[i]["artist_id"] == artist_id) {
                jsonData[i]["path_lengths"] = pathLengths;
                jsonData[i]["time"] = execTime;
                break;
            }
        }

        ofstream outFile(fileName);
        if (!outFile.is_open()) {
            throw std::runtime_error("Could not open output file.");
        }
        outFile << jsonData.dump(4); // Write jsonData to output file
        outFile.close();
    }
    catch (const exception& e) {
        cerr << "Error: " << e.what() << std::endl;

    }
}

string getUnsearchedArtist(){
    try{
        const string fileName = "data/1000_random_artists_cpp.json";
    
        ifstream inFile(fileName);
        if (!inFile.is_open()) {
            throw std::runtime_error("Could not open input file.");
        }
        nlohmann::json jsonData;
        inFile >> jsonData; // Parse JSON file content into jsonData
        inFile.close();

        for (int i = 0; i < jsonData.size(); i++){
            if (jsonData[i]["path_lengths"] == 0) {
                cout << "Num of searched: " << i << endl;
                return jsonData[i]["artist_id"];
            }
        }
        cout << "none" << endl;
        return "";

    }
    catch (const exception& e) {
        cerr << "Error: " << e.what() << std::endl;
        return "";
    }
    return "";
}

 int getLenghts(string artist_id, sqlite3* db, unordered_map<int, int>& pathLengths){
    auto startTime = chrono::high_resolution_clock::now();
    unordered_set<string> explored;
    deque<vector<string>> queue;
    vector<string> path;
    vector<string> newPath;
    string node;
    vector<string> relatedArtists;

    queue.push_back({artist_id});
    explored.insert(artist_id);
    
    while (queue.size() > 0){
        
        path = queue.front();
        queue.pop_front();
        node = path.back(); 

        relatedArtists = getRelatedArtists(node, db);
        for (string relatedArtist: relatedArtists){
            if(explored.find(relatedArtist) == explored.end()){
                explored.insert(relatedArtist);
                newPath = path;
                newPath.push_back(relatedArtist);
                queue.push_back(newPath);
            }
        }
        
        pathLengths[path.size()-1] = pathLengths[path.size()-1] + 1;

    }

    auto end = chrono::high_resolution_clock::now();
    auto execTime = chrono::duration_cast<chrono::seconds>(end - startTime);
    cout << "Execution time: " << execTime.count() << " seconds" << endl;

    return execTime.count();
}

int findShortestPath(){
    sqlite3* db;
    if (sqlite3_open("../artists.db", &db) != SQLITE_OK) {
        cerr << "Error opening database: " << sqlite3_errmsg(db) << endl;
        return 1;
    }
    cout << "Database opened successfully!" << endl;
    while (true){
        string artist_id = getUnsearchedArtist();
        if (artist_id == ""){
            break;
        }
        // Key: Length of a path, Value: Number of paths with that length
        unordered_map<int, int> pathLengths;
        int execTime = getLenghts(artist_id, db, pathLengths);
        writeToJSON(artist_id, pathLengths, execTime);

    }    
    sqlite3_close(db); 
    return 0;

}

File: data_analysis/find_shortest_path.py
from handlers.db_handler import *
import sqlite3
from handlers.spotify_api_handler import get_artist_spotify, get_token
from data_analysis.search_methods import bfs, dfs, get_all_lengths
from datetime import datetime
import json



def get_all_lengths(a1, cursor):
    """
    Find the path length from one artist to all others in the database using breadth-first search
    ##Write to a json file
    :param a1: The id of the artist
    :param cursor: The cursor to the database
    :return: Dict of paths lengths
    """
    start_time = datetime.now().timestamp()
    
    explored = {}
    found = {a1: None}
    # Key = path length, Value = number of artists with that path length
    path_lengths = {}
    queue = [[a1]]
    while queue:
        path = queue.pop(0)
        node = path[-1]

        for related_artist in get_related_artists(node, cursor):
            if related_artist[0] not in found:
                found[related_artist[0]] = None
                new_path = list(path)
                new_path.append(related_artist[0])
                queue.append(new_path)

        explored[node] = None
        path_lengths[len(path)-1] = path_lengths.get(len(path) -1, 0) + 1
        #print("Path lengths: ", path_lengths, " Queue length: ", len(queue), end="\r")
    artists_found = 0
    for i in range(len(path_lengths)):
        artists_found += path_lengths[i]
    if artists_found == len(get_all("all_artists", cursor)):
        print("\nAll artists found")
    else:
        print("\nFound: ", artists_found, " artists of ", len(get_all("all_artists", cursor)))
    exec_time = datetime.now().timestamp() - start_time
    print("Time: ", exec_time)
    return path_lengths, round(exec_time, 2)


def create_1000_random_artists_file(cursor):
    """
    ONLY RUN ONCE
    """
    json_list = []
    artists_list = get_1000_random_artists(cursor)
    for artist in artists_list:
        json_list.append({"artist_id": artist[0], "path_lengths": 0, "time": 0})
    with open('data/1000_random_artists_python.json', 'w') as f:
        json.dump(json_list, f, indent=4)
    with open('data/1000_random_artists_cpp.json', 'w') as f:
        json.dump(json_list, f, indent=4)


def write_to_json_file(artist_id, path_lengths, exec_time):
    with open("data/1000_random_artists_python.json", 'r') as f:
        data = json.load(f)
    for artist in data:
        if artist["artist_id"] == artist_id:
            artist["path_lengths"] = path_lengths
            artist["time"] = exec_time
            break
    with open("data/1000_random_artists_python.json", 'w') as f:
        json.dump(data, f, indent=4)


def get_unsearched_artist():
    with open("data/1000_random_artists_python.json", 'r') as f:
        data = json.load(f)
    for artist in data:
        if artist["path_lengths"] == 0:
            return artist["artist_id"]
    return None

        





        

def main():
    conn = sqlite3.connect('../artists.db')
    cursor = conn.cursor()
    start_time = datetime.now().timestamp()

    while True:
        artist_id = get_unsearched_artist()
        if artist_id is None:
            break
        path_lengths, exec_time = get_all_lengths(artist_id, cursor)
        write_to_json_file(artist_id, path_lengths, exec_time)
        print("Time: ", datetime.now().timestamp() - start_time)

    

    conn.close()